src/mame/drivers/firebeat.cpp

```diff
@@ -1235,315 +1546,439 @@ void firebeat_state::firebeat2(machine_config &config)
+uint8_t firebeat_kbm_state::midi_uart_r(offs_t offset)
+{
+	return m_duart_midi->read(offset >> 6);
+}
 
-	rf5c400_device &rf5c400(RF5C400(config, "rf5c400", XTAL(16'934'400)));
-	rf5c400.set_addrmap(0, &firebeat_state::rf5c400_map);
-	rf5c400.add_route(0, "lspeaker", 0.5);
-	rf5c400.add_route(1, "rspeaker", 0.5);
+void firebeat_kbm_state::midi_uart_w(offs_t offset, uint8_t data)
+{
+	m_duart_midi->write(offset >> 6, data);
 }
 
-void firebeat_state::firebeat_spu(machine_config &config)
+WRITE_LINE_MEMBER(firebeat_kbm_state::midi_uart_ch0_irq_callback)
 {
-	firebeat_spu_base(config);
+	if (BIT(m_extend_board_irq_enable, 1) == 0 && state != CLEAR_LINE)
+	{
+		m_extend_board_irq_active |= 0x02;
+		m_maincpu->set_input_line(INPUT_LINE_IRQ1, ASSERT_LINE);
+	}
+	else
+		m_maincpu->set_input_line(INPUT_LINE_IRQ1, CLEAR_LINE);
+}
 
-	ATA_INTERFACE(config, m_spuata).options(firebeat_ata_devices, "cdrom", nullptr, true);
-	m_spuata->slot(0).set_option_machine_config("cdrom", dvdrom_config);
-	m_spuata->irq_handler().set(FUNC(firebeat_state::spu_ata_interrupt));
-	m_spuata->dmarq_handler().set(FUNC(firebeat_state::spu_ata_dmarq));
+WRITE_LINE_MEMBER(firebeat_kbm_state::midi_uart_ch1_irq_callback)
+{
+	if (BIT(m_extend_board_irq_enable, 0) == 0 && state != CLEAR_LINE)
+	{
+		m_extend_board_irq_active |= 0x01;
+		m_maincpu->set_input_line(INPUT_LINE_IRQ1, ASSERT_LINE);
+	}
+	else
+		m_maincpu->set_input_line(INPUT_LINE_IRQ1, CLEAR_LINE);
 }
 
-void firebeat_state::firebeat_spu_bm3(machine_config &config)
+/*
+static const int keyboard_notes[24] =
 {
-	firebeat_spu_base(config);
+    0x3c,   // C1
+    0x3d,   // C1#
+    0x3e,   // D1
+    0x3f,   // D1#
+    0x40,   // E1
+    0x41,   // F1
+    0x42,   // F1#
+    0x43,   // G1
+    0x44,   // G1#
+    0x45,   // A1
+    0x46,   // A1#
+    0x47,   // B1
+    0x48,   // C2
+    0x49,   // C2#
+    0x4a,   // D2
+    0x4b,   // D2#
+    0x4c,   // E2
+    0x4d,   // F2
+    0x4e,   // F2#
+    0x4f,   // G2
+    0x50,   // G2#
+    0x51,   // A2
+    0x52,   // A2#
+    0x53,   // B2
+};
 
-	screen_device *screen = subdevice<screen_device>("screen");
-	if (screen != nullptr) {
-		// beatmania III is the only game on the Firebeat platform to use 640x480
-		screen->set_size(640, 480);
-		screen->set_visarea(0, 639, 0, 479);
-	}
+TIMER_CALLBACK_MEMBER(firebeat_kbm_state::keyboard_timer_callback)
+{
+    static const int kb_uart_channel[2] = { 1, 0 };
+    static const char *const keynames[] = { "KEYBOARD_P1", "KEYBOARD_P2" };
+    int keyboard;
+    int i;
 
-	ATA_INTERFACE(config, m_spuata).options(firebeat_ata_devices, "hdd", nullptr, true);
-	m_spuata->irq_handler().set(FUNC(firebeat_state::spu_ata_interrupt));
-	m_spuata->dmarq_handler().set(FUNC(firebeat_state::spu_ata_dmarq));
-}
+    for (keyboard=0; keyboard < 2; keyboard++)
+    {
+        uint32_t kbstate = ioport(keynames[keyboard])->read();
```

Comment:
This is still a tag map lookup here.

Reply:
Actually, is it inside a comment?  The diff view is messing with me, I can barely tell.

Reply:
Yeah, this is in a comment. I left it in because it has a some code for Keyboardmania's MIDI keyboard implementation which I think may be useful if someone wants to go down that path in the future.